pkg/proxy/proxy.go

```diff
@@ -155,33 +157,41 @@ func (p *Proxy) health(ctx echo.Context) error {
 }
 
 func (p *Proxy) handleRequestAndRedirect(ctx echo.Context) error {
+	startTime := ctx.Get(context.StartTime).(time.Time)
 	userID, _ := ctx.Get(context.SubKey).(string)
 	username, _ := ctx.Get(context.UsernameKey).(string)
 
 	proxyPluginName, workspace, err := getWorkspaceContext(ctx.Request())
 	if err != nil {
+		metrics.RegServProxyResponseHistogramVec.WithLabelValues(metrics.ResponseMetricLabelReject).Observe(time.Since(startTime).Seconds())
 		return crterrors.NewBadRequest("unable to get workspace context", err.Error())
 	}
 	ctx.Set(context.WorkspaceKey, workspace) // set workspace context for logging
 
 	cluster, err := p.app.MemberClusterService().GetClusterAccess(userID, username, workspace, proxyPluginName)
 	if err != nil {
+		metrics.RegServProxyResponseHistogramVec.WithLabelValues(metrics.ResponseMetricLabelReject).Observe(time.Since(startTime).Seconds())
 		return crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
 	}
 
 	// before proxying the request, verify that the user has a spacebinding for the workspace and that the namespace (if any) belongs to the workspace
 	workspaces, err := p.spaceLister.ListUserWorkspaces(ctx)
 	if err != nil {
+		metrics.RegServProxyResponseHistogramVec.WithLabelValues(metrics.ResponseMetricLabelReject).Observe(time.Since(startTime).Seconds())
 		return crterrors.NewInternalError(errs.New("unable to retrieve user workspaces"), err.Error())
 	}
 
 	requestedNamespace := namespaceFromCtx(ctx)
 	if err := validateWorkspaceRequest(workspace, requestedNamespace, workspaces); err != nil {
+		metrics.RegServProxyResponseHistogramVec.WithLabelValues(metrics.ResponseMetricLabelReject).Observe(time.Since(startTime).Seconds())
 		return crterrors.NewForbiddenError("invalid workspace request", err.Error())
 	}
```

Comment:
there are many duplication of the same for every error. It's just a suggestion, but we could extract that part to a separate method and calculate the time only once:
```go
func (p *Proxy) processRequest(ctx echo.Context) (string, *access.ClusterAccess, error) {
	userID, _ := ctx.Get(context.SubKey).(string)
	username, _ := ctx.Get(context.UsernameKey).(string)

	proxyPluginName, workspace, err := getWorkspaceContext(ctx.Request())
	if err != nil {
		return "", nil, crterrors.NewBadRequest("unable to get workspace context", err.Error())
	}
	ctx.Set(context.WorkspaceKey, workspace) // set workspace context for logging

	cluster, err := p.app.MemberClusterService().GetClusterAccess(userID, username, workspace, proxyPluginName)
	if err != nil {
		return "", nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
	}

	// before proxying the request, verify that the user has a spacebinding for the workspace and that the namespace (if any) belongs to the workspace
	workspaces, err := p.spaceLister.ListUserWorkspaces(ctx)
	if err != nil {
		return "", nil, crterrors.NewInternalError(errs.New("unable to retrieve user workspaces"), err.Error())
	}

	requestedNamespace := namespaceFromCtx(ctx)
	if err := validateWorkspaceRequest(workspace, requestedNamespace, workspaces); err != nil {
		return "", nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
	}
	return proxyPluginName, cluster, nil
}

func (p *Proxy) handleRequestAndRedirect(ctx echo.Context) error {
	startTime := ctx.Get(context.StartTime).(time.Time)

	proxyPluginName, cluster, err := p.processRequest(ctx)
	if err != nil {
		metrics.RegServProxyResponseHistogramVec.WithLabelValues(metrics.ResponseMetricLabelReject).Observe(time.Since(startTime).Seconds())
		return err
	}

 ...
```